### PR TITLE
feat: introduce blockstore manager

### PR DESCRIPTION
**Pull request by emizzle**
Implement blockstore manager which executes block storage operations on its block stores, in the order to which they were added to the manager, typically in the order of most local (fastest, eg cache) to least local (slowest, eg filesystem or perhaps a network filesystem). As an example, given a `BlockStoreManager` instantiated with a `@[MemoryStore, FSStore]`, retrieving a block would first attempt to get from the `MemoryStore`, and if not found, attempt to get from the `FSStore`.

Remove all dependencies on `BlockStores` (typically in the shape of `localstore`) and instead depend on `BlockStoreManager` via the `BlockExcEngine`.

Modify the role of the `BlockExcEngine` to make a “local vs remote” decision on block access/storage. For all operations other than retrieving blocks, this means simply going to the `BlockStoreManager`. For retrieving blocks, however, this means going first to the `BlockStoreManager`, and then if not found, going to the Dagger network (via pending block and want/have lists).

Remove `NetworkStore` as its two purposes were to defer block retrieval from a local store first, then go to the block exchange to requeest a block from the Dagger network. `BlockStoreManager` takes care of going to local storage first, and the block exchange engine handles going to Dagger network if retrieval from the store manager fails.

### Notes
1. Future work may want to consider breaking up `BlockExcEngine` further in to three modules:
  - `BlockExcEngine` (depends on `WantHave`, `DHT`)
  - `WantHave`
  - `DHT` (work is in progress)

Co-authored-by: Michael Bradley <michaelsbradleyjr@gmail.com>

### CODE REVIEW

**Review by dryajov — CHANGES_REQUESTED** [review]

**Comment by dryajov** [comment]
Great job! But I'm not entirely sure about the removal of the network store, which is the reason why I eventually scrapped the manager idea some time ago. The reason is because there is logic around which store needs to be accessed first and this gets a bit messy with the flat structure of the manager. We can still accomplish that with the manager, but the logic for which store to access when, would need to migrate to the node (as I pointed out in the review comment), in other words it wouldn't be as self contained as in the current implementation.

This isn't a big deal, but it does leave us with two entities to work with instead of one. The way I see it is as this:

- The original implementation relies on the stores allows us to switch different implementations seamlessly since all the logic for deciding which store to access is encapsulated in the stores themself. For example, the current implementation simply passes the fs and memory stores to the network store and the network store decides which one needs to be accessed and how.
- The new implementation introduces a manager that will hold different stores and access them in sequence in the order they where registered and separately relies on the engine to perform all the network duties. This makes the blockexchange engine a primary abstraction that needs to be carried around in addition to the block stores. I'm not sure if this brings any benefits.

In other words, the block stores was chosen as an abstraction because it allows us to neatly incapsulate most of the functionality in the stores themself and allows switching different pieces independently. For example, if we want to have several concurrent network stores for different use cases, we can do that without much changes, since the logic for the new network store would be encapsulated behind the same interface. 

I like the manager and I think we can keep it, but I would rather we keep the network store as in the current implementation and only use the manager as an abstraction to switch between different concurrent network stores, for example.

**Comment by emizzle** [comment]
Thanks so much for the feedback @dryajov. While Michael and I formulate some clarifications and questions, could you please explain what you mean by `switch between different concurrent network stores`?

File: dagger/node.nim [hunk 1]
```diff
@@ -71,7 +70,7 @@ proc streamBlocks*(
     # to prevent slurping the entire dataset
     # since disk IO is blocking
     for c in blockManifest:
-      without blk =? (await node.blockStore.getBlock(c)):
+      without blk =? (await node.engine.retrieve(c)):
```
[the remark this answers is not in the document]
Comment (michaelsbradleyjr) [reply dagger/node.nim:73]: In this PR, calling `node.engine.retrieve(c)` results in the engine first calling the block store manager, which checks if the block is available in any of the local block stores, e.g. the cache or file system store. If it’s available locally, then it will be returned. If it’s not available locally, the engine will attempt to get the block from the p2p network.

On one hand we have block stores which are resources managed by their nodes, and on the other we have the p2p network, i.e. the Dagger (Codex) Network. The main point being that a block store is a storage resource that logically sits “behind” a node. This represents a slight shift in terminology, where “block store” implies a storage resource that is (relatively) local to the node. Again that could be an in-memory cache or the local file system, but it could also be a NAS or S3 bucket or something like that.

By separating the concepts of block *store* and block exchange *network*, which in this PR involves dropping the idea/type `NetworkStore`, @emizzle and I were hoping to:

1. cut down on confusion when thinking about and discussing concepts related to “network”
2. remove a tight coupling of modules between `BlockExcEngine` and `NetworkStore`, where the only function of `NetworkStore` outside of the functionality covered by block store manager was to call the engine.

**Comment by michaelsbradleyjr** [comment]
> We can still accomplish that with the manager, but the logic for which store to access when, would need to migrate to the node (as I pointed out in the review comment), in other words it wouldn't be as self contained as in the current implementation.

See the [reply above](https://github.com/status-im/nim-dagger/pull/49#discussion_r807410062). Quick recap: the node calls the engine, which first tries to get the block from the block store manager. The manager is responsible for trying to get the block from any of the node’s storage resources, i.e. its block stores. If the manager doesn’t return the block then the engine tries to get the block from the p2p network.

The node could query the manager instead of the engine doing so, but on first pass the engine seemed to be the more natural place for that query, given that there was already logic in the engine that directly accesses `localStore`, e.g. in `scheduleTasks`. All references to `localStore` within the engine were changed to `blockStoreMgr`, so it also seemed natural to encapsulate the logic of *“first try the manager, then go to the network”* in the engine.

An idea that @emizzle and I discussed yesterday is to split out the block exchange logic from the engine, so the engine becomes more of a pure coordination layer. See the diagram below for a rough sketch. This PR could be seen as a step in that direction.

Codex_layers2

**Comment by emizzle** [comment]
After a lengthy conversation, we are closing this as the abstraction needs to happen at the engine/block exchange layer, such that there may be multiple parallel engines. The blockstore manager approach assumes the blockexchange is a singular instance. The babushka doll approach to the stores remains.
